File: src/csharp/Foundgine.Providers/Storage/Sql/Retrieval/PostgresRetrievalCandidateSource.cs
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Metadata;
using Foundgine.Core.Semantic.Resolution;
using Npgsql;

namespace Foundgine.Providers.Storage.Sql.Retrieval;

/// <summary>
/// PostgreSQL implementation of the semantic candidate boundary.
///
/// This provider generates retrieval candidates and evidence only.
/// Authorization and final relational execution remain owned by
/// Foundgine's semantic/execution pipeline.
/// </summary>
public sealed class PostgresRetrievalCandidateSource
    : IApproximateCandidateSource
{
    private readonly NpgsqlDataSource _dataSource;

    private readonly IMetadataCatalog _metadata;

    private readonly PostgresRetrievalOptions _options;

    public PostgresRetrievalCandidateSource(
        NpgsqlDataSource dataSource,
        IMetadataCatalog metadata,
        PostgresRetrievalOptions? options = null)
    {
        _dataSource =
            dataSource ??
            throw new ArgumentNullException(nameof(dataSource));

        _metadata =
            metadata ??
            throw new ArgumentNullException(nameof(metadata));

        _options = options ?? new();
    }

    public IReadOnlyList<RetrievalCandidate> Retrieve(
        SemanticRetrievalRequest request)
    {
        return RetrieveAsync(request)
            .GetAwaiter()
            .GetResult();
    }

    public async Task<IReadOnlyList<RetrievalCandidate>> RetrieveAsync(
        SemanticRetrievalRequest request,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);

        var entity =
            _metadata.GetEntity(request.EntityType);

        var field =
            ResolveField(entity, request.Field);

        return request.Strategy switch
        {
            RetrievalStrategy.Fuzzy =>
                _options.EnablePgTrgm
                    ? await ExecuteFuzzyAsync(
                        entity,
                        field,
                        request,
                        cancellationToken)
                    : throw Unsupported(
                        RetrievalStrategy.Fuzzy),

            RetrievalStrategy.FullText =>
                _options.EnableFullText
                    ? await ExecuteFullTextAsync(
                        entity,
                        field,
                        request,
                        cancellationToken)
                    : throw Unsupported(
                        RetrievalStrategy.FullText),

            RetrievalStrategy.Search =>
                _options.EnablePgSearch
                    ? await ExecutePgSearchAsync(
                        entity,
                        field,
                        request,
                        cancellationToken)
                    : throw Unsupported(
                        RetrievalStrategy.Search),

            RetrievalStrategy.GraphSimilarity =>
                _options.EnableApacheAge
                    ? await ExecuteGraphSimilarityAsync(
                        entity,
                        request,
                        cancellationToken)
                    : throw Unsupported(
                        RetrievalStrategy.GraphSimilarity),

            RetrievalStrategy.Vector =>
                throw new NotSupportedException(
                    "This field-value retrieval boundary does not implement " +
                    "vector search. For token-level lexical grounding backed " +
                    "by pgvector, use PgVectorSemanticLexicalCandidateSource " +
                    "in Foundgine.Providers.Storage.PostgresVector, which implements " +
                    "ISemanticLexicalCandidateSource directly against the " +
                    "projected semantic lexicon."),

            RetrievalStrategy.Relational =>
                [],

            _ =>
                throw new ArgumentOutOfRangeException(
                    nameof(request.Strategy),
                    request.Strategy,
                    "Unknown retrieval strategy.")
        };
    }

    private async Task<IReadOnlyList<RetrievalCandidate>>
        ExecuteFuzzyAsync(
            EntityMetadata entity,
            FieldMetadata field,
            SemanticRetrievalRequest request,
            CancellationToken cancellationToken)
    {
        var table =
            Quote(entity.EffectiveStorageName);

        var column =
            Quote(GetColumnName(entity, field));

        var identity =
            Quote(GetPrimaryKeyName(entity));

        var sql =
            $"""
             SELECT
                 {identity},
                 similarity({column}, $1) AS score
             FROM {table}
             WHERE {column} % $1
             ORDER BY score DESC
             LIMIT $2
             """;

        await using var command =
            _dataSource.CreateCommand(sql);

        command.Parameters.AddWithValue(
            request.Query);

        command.Parameters.AddWithValue(
            request.Limit);

        return await ReadCandidatesAsync(
            command,
            entity.EntityId,
            field.Id,
            CandidateEvidenceKind.Trigram,
            cancellationToken);
    }

    private async Task<IReadOnlyList<RetrievalCandidate>>
        ExecuteFullTextAsync(
            EntityMetadata entity,
            FieldMetadata field,
            SemanticRetrievalRequest request,
            CancellationToken cancellationToken)
    {
        var table =
            Quote(entity.EffectiveStorageName);

        var column =
            Quote(GetColumnName(entity, field));

        var identity =
            Quote(GetPrimaryKeyName(entity));

        var config =
            QuoteLiteral(
                _options.FullTextConfiguration);

        var sql =
            $"""
             SELECT
                 {identity},
                 ts_rank_cd(
                     to_tsvector(
                         {config},
                         COALESCE({column}::text, '')
                     ),
                     websearch_to_tsquery(
                         {config},
                         $1
                     )
                 ) AS score
             FROM {table}
             WHERE to_tsvector(
                 {config},
                 COALESCE({column}::text, '')
             ) @@ websearch_to_tsquery(
                 {config},
                 $1
             )
             ORDER BY score DESC
             LIMIT $2
             """;

        await using var command =
            _dataSource.CreateCommand(sql);

        command.Parameters.AddWithValue(
            request.Query);

        command.Parameters.AddWithValue(
            request.Limit);

        return await ReadCandidatesAsync(
            command,
            entity.EntityId,
            field.Id,
            CandidateEvidenceKind.FullText,
            cancellationToken);
    }

    private async Task<IReadOnlyList<RetrievalCandidate>>
        ExecutePgSearchAsync(
            EntityMetadata entity,
            FieldMetadata field,
            SemanticRetrievalRequest request,
            CancellationToken cancellationToken)
    {
        /*
         * pg_search is deliberately isolated to the PostgreSQL provider.
         *
         * The semantic layer knows only about:
         *
         *     RetrievalStrategy.Search
         *
         * It does not know that PostgreSQL happens to implement that
         * strategy using pg_search/BM25.
         *
         * This keeps the semantic contracts provider-neutral.
         */

        var table =
            Quote(entity.EffectiveStorageName);

        var column =
            Quote(GetColumnName(entity, field));

        var identity =
            Quote(GetPrimaryKeyName(entity));

        var sql =
            $"""
             SELECT
                 {identity},
                 pdb.score({identity}) AS score
             FROM {table}
             WHERE {column} ||| $1
             ORDER BY score DESC
             LIMIT $2
             """;

        await using var command =
            _dataSource.CreateCommand(sql);

        command.Parameters.AddWithValue(
            request.Query);

        command.Parameters.AddWithValue(
            request.Limit);

        return await ReadCandidatesAsync(
            command,
            entity.EntityId,
            field.Id,
            CandidateEvidenceKind.Bm25,
            cancellationToken);
    }

    private async Task<IReadOnlyList<RetrievalCandidate>>
        ExecuteGraphSimilarityAsync(
            EntityMetadata entity,
            SemanticRetrievalRequest request,
            CancellationToken cancellationToken)
    {
        if (request.Relationship is null)
        {
            throw new ArgumentException(
                "GraphSimilarity requires Relationship.",
                nameof(request));
        }

        if (request.ReferenceIdentity is null)
        {
            throw new ArgumentException(
                "GraphSimilarity requires ReferenceIdentity.",
                nameof(request));
        }

        var relationship =
            _metadata.GetRelationship(
                request.Relationship.Value);

        var graph =
            _options.AgeGraphName;

        var sourceLabel =
            QuoteIdentifier(
                _metadata
                    .GetEntity(relationship.Source)
                    .EffectiveStorageName);

        var targetLabel =
            QuoteIdentifier(
                _metadata
                    .GetEntity(relationship.Target)
                    .EffectiveStorageName);

        var edgeLabel =
            QuoteIdentifier(
                relationship.Name);

        var reference =
            request.ReferenceIdentity
                .Replace("\\", "\\\\")
                .Replace("\"", "\\\"");

        var cypher =
            $"""
             MATCH
                 (reference:{sourceLabel})
                 -[:{edgeLabel}]->
                 (neighbor)
                 <-[:{edgeLabel}]-
                 (candidate:{sourceLabel})
             WHERE
                 reference.id = "{reference}"
                 AND candidate.id <> "{reference}"
             RETURN
                 candidate.id,
                 count(*) AS score
             ORDER BY score DESC
             LIMIT {request.Limit}
             """;

        await using var load =
            _dataSource.CreateCommand(
                "LOAD 'age'");

        await load.ExecuteNonQueryAsync(
            cancellationToken);

        var sql =
            $"""
             SELECT *
             FROM cypher(
                 {QuoteLiteral(graph)},
                 $1
             ) AS (
                 record_id text,
                 score bigint
             )
             """;

        await using var command =
            _dataSource.CreateCommand(sql);

        command.Parameters.AddWithValue(
            cypher);

        return await ReadCandidatesAsync(
            command,
            entity.EntityId,
            null,
            CandidateEvidenceKind.GraphSimilarity,
            cancellationToken);
    }

    private static async Task<IReadOnlyList<RetrievalCandidate>>
        ReadCandidatesAsync(
            NpgsqlCommand command,
            EntityId entityType,
            FieldId? field,
            CandidateEvidenceKind kind,
            CancellationToken cancellationToken)
    {
        var results =
            new List<RetrievalCandidate>();

        await using var reader =
            await command.ExecuteReaderAsync(
                cancellationToken);

        while (await reader.ReadAsync(
                   cancellationToken))
        {
            var recordId =
                Convert.ToString(
                    reader.GetValue(0),
                    System.Globalization.CultureInfo.InvariantCulture)
                ?? string.Empty;

            var score =
                reader.IsDBNull(1)
                    ? 0d
                    : Convert.ToDouble(
                        reader.GetValue(1),
                        System.Globalization.CultureInfo.InvariantCulture);

            results.Add(
                new RetrievalCandidate(
                    entityType,
                    recordId,
                    score,
                    field,
                    recordId,
                    [
                        new ResolutionEvidence(
                            $"{kind} retrieval matched semantic " +
                            $"candidate {recordId} with score " +
                            $"{score:0.####}.",
                            kind,
                            score)
                    ],
                    kind));
        }

        return results;
    }

    private FieldMetadata ResolveField(
        EntityMetadata entity,
        FieldId? fieldId)
    {
        if (fieldId is { } id)
        {
            return entity.EffectiveFields
                       .FirstOrDefault(x => x.Id == id)
                   ?? throw new KeyNotFoundException(
                       $"Field {id} is not registered " +
                       $"on entity {entity.Name}.");
        }

        return entity.EffectiveFields
                   .FirstOrDefault(x =>
                       x.ClrType == typeof(string) &&
                       x.Column is not null)
               ?? throw new InvalidOperationException(
                   $"Entity {entity.Name} has no string field " +
                   $"available for approximate retrieval.");
    }

    private string GetColumnName(
        EntityMetadata entity,
        FieldMetadata field)
    {
        if (field.Column is null)
        {
            throw new InvalidOperationException(
                $"Field {field.Name} has no storage column.");
        }

        return entity.Columns
                   .FirstOrDefault(x => x.Id == field.Column.ColumnId)
                   ?.EffectiveStorageName
               ?? throw new InvalidOperationException(
                   $"Column {field.Column.ColumnId} for field " +
                   $"{field.Name} is not registered on entity " +
                   $"{entity.Name}.");
    }

    private string GetPrimaryKeyName(
        EntityMetadata entity)
    {
        if (entity.PrimaryKey is null)
        {
            throw new InvalidOperationException(
                $"Entity {entity.Name} has no primary key metadata.");
        }

        return entity.Columns
                   .FirstOrDefault(x => x.Id == entity.PrimaryKey.ColumnId)
                   ?.EffectiveStorageName
               ?? throw new InvalidOperationException(
                   $"Primary key column " +
                   $"{entity.PrimaryKey.ColumnId} is not registered " +
                   $"on entity {entity.Name}.");
    }

    private static string Quote(
        string value)
    {
        return QuoteIdentifier(value);
    }

    private static string QuoteIdentifier(
        string value)
    {
        return "\"" +
               value.Replace("\"", "\"\"") +
               "\"";
    }

    private static string QuoteLiteral(
        string value)
    {
        return "'" +
               value.Replace("'", "''") +
               "'";
    }

    private static NotSupportedException Unsupported(
        RetrievalStrategy strategy)
    {
        return new NotSupportedException(
            $"PostgreSQL retrieval strategy " +
            $"{strategy} is disabled.");
    }
}

File: src/java/foundgine-providers/src/main/java/com/foundgine/providers/storage/sql/retrieval/PostgresRetrievalCandidateSource.java
package com.foundgine.providers.storage.sql.retrieval;

import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.metadata.*;
import com.foundgine.core.semantic.resolution.*;

import javax.sql.DataSource;
import java.sql.*;
import java.util.*;

/**
 * PostgreSQL implementation of the semantic candidate boundary.
 *
 * <p>
 * Retrieval is advisory evidence only. Semantic resolution and authorization
 * remain outside the physical retrieval provider.
 * </p>
 */
public final class PostgresRetrievalCandidateSource implements IApproximateCandidateSource {
	private final DataSource dataSource;
	private final IMetadataCatalog metadata;
	private final PostgresRetrievalOptions options;

	public PostgresRetrievalCandidateSource(DataSource dataSource, IMetadataCatalog metadata) {
		this(dataSource, metadata, new PostgresRetrievalOptions());
	}

	public PostgresRetrievalCandidateSource(DataSource dataSource, IMetadataCatalog metadata,
			PostgresRetrievalOptions options) {
		this.dataSource = Objects.requireNonNull(dataSource, "dataSource");
		this.metadata = Objects.requireNonNull(metadata, "metadata");
		this.options = options == null ? new PostgresRetrievalOptions() : options;
	}

	@Override
	public List<RetrievalCandidate> retrieve(SemanticRetrievalRequest request) {
		Objects.requireNonNull(request, "request");
		EntityMetadata entity = metadata.getEntity(request.entityType());

		return switch (request.strategy()) {
		case FUZZY -> {
			requireEnabled(options.enablePgTrgm(), RetrievalStrategy.FUZZY);
			yield executeText(entity, resolveField(entity, request.field()), request, CandidateEvidenceKind.TRIGRAM,
					fuzzySql(entity, resolveField(entity, request.field())));
		}
		case FULL_TEXT -> {
			requireEnabled(options.enableFullText(), RetrievalStrategy.FULL_TEXT);
			yield executeText(entity, resolveField(entity, request.field()), request, CandidateEvidenceKind.FULL_TEXT,
					fullTextSql(entity, resolveField(entity, request.field())));
		}
		case SEARCH -> {
			requireEnabled(options.enablePgSearch(), RetrievalStrategy.SEARCH);
			yield executeText(entity, resolveField(entity, request.field()), request, CandidateEvidenceKind.BM25,
					searchSql(entity, resolveField(entity, request.field())));
		}
		case GRAPH_SIMILARITY -> {
			requireEnabled(options.enableApacheAge(), RetrievalStrategy.GRAPH_SIMILARITY);
			yield executeGraphSimilarity(entity, request);
		}
		case RELATIONAL -> List.of();
		case VECTOR -> throw new UnsupportedOperationException(
				"This field-value retrieval boundary does not implement vector search. "
						+ "Use the pgvector semantic lexical candidate source instead.");
		default -> throw new IllegalArgumentException("Unknown retrieval strategy: " + request.strategy());
		};
	}

	private List<RetrievalCandidate> executeText(EntityMetadata entity, FieldMetadata field,
			SemanticRetrievalRequest request, CandidateEvidenceKind evidenceKind, String sql) {
		try (Connection connection = dataSource.getConnection();
				PreparedStatement statement = connection.prepareStatement(sql)) {
			int i = 1;
			statement.setString(i++, request.query());
			if (request.strategy() == RetrievalStrategy.FUZZY || request.strategy() == RetrievalStrategy.FULL_TEXT)
				statement.setString(i++, request.query());
			statement.setInt(i, request.limit());
			try (ResultSet result = statement.executeQuery()) {
				return readCandidates(result, entity.entityId(), field.id(), evidenceKind);
			}
		} catch (SQLException ex) {
			throw new IllegalStateException("PostgreSQL " + request.strategy() + " retrieval failed.", ex);
		}
	}

	private List<RetrievalCandidate> executeGraphSimilarity(EntityMetadata entity, SemanticRetrievalRequest request) {
		if (request.relationship() == null)
			throw new IllegalArgumentException("GraphSimilarity requires Relationship.");
		if (request.referenceIdentity() == null || request.referenceIdentity().isBlank())
			throw new IllegalArgumentException("GraphSimilarity requires ReferenceIdentity.");

		RelationshipMetadata relationship = metadata.getRelationship(request.relationship());
		EntityMetadata source = metadata.getEntity(relationship.source());
		EntityMetadata target = metadata.getEntity(relationship.target());
		String sourceLabel = quoteIdentifier(source.effectiveStorageName());
		String edgeLabel = quoteIdentifier(relationship.name());
		String reference = request.referenceIdentity().replace("\\", "\\\\").replace("\"", "\\\"");
		String cypher = "MATCH (reference:" + sourceLabel + ")-[:" + edgeLabel + "]->(neighbor) " + "<-[:" + edgeLabel
				+ "]-(candidate:" + sourceLabel + ") " + "WHERE reference.id = \"" + reference
				+ "\" AND candidate.id <> \"" + reference + "\" " + "RETURN candidate.id, count(*) AS score "
				+ "ORDER BY score DESC LIMIT " + request.limit();

		String sql = "SELECT * FROM cypher(" + quoteLiteral(options.ageGraphName()) + ", ?) "
				+ "AS (record_id text, score bigint)";
		try (Connection connection = dataSource.getConnection(); Statement load = connection.createStatement()) {
			load.execute("LOAD 'age'");
			try (PreparedStatement command = connection.prepareStatement(sql)) {
				command.setString(1, cypher);
				try (ResultSet result = command.executeQuery()) {
					return readCandidates(result, entity.entityId(), null, CandidateEvidenceKind.GRAPH_SIMILARITY);
				}
			}
		} catch (SQLException ex) {
			throw new IllegalStateException("PostgreSQL GraphSimilarity retrieval failed.", ex);
		}
	}

	private static List<RetrievalCandidate> readCandidates(ResultSet result, EntityId entityType, FieldId field,
			CandidateEvidenceKind kind) throws SQLException {
		List<RetrievalCandidate> candidates = new ArrayList<>();
		while (result.next()) {
			String recordId = Objects.toString(result.getObject(1), "");
			double score = result.getObject(2) == null ? 0d : result.getDouble(2);
			candidates.add(new RetrievalCandidate(entityType, recordId, score, field, recordId,
					List.of(new ResolutionEvidence(kind + " retrieval matched semantic candidate " + recordId
							+ " with score " + formatScore(score) + ".", kind, score)),
					kind));
		}
		return List.copyOf(candidates);
	}

	private FieldMetadata resolveField(EntityMetadata entity, FieldId fieldId) {
		if (fieldId != null) {
			return entity.effectiveFields().stream().filter(x -> x.id().equals(fieldId)).findFirst()
					.orElseThrow(() -> new NoSuchElementException(
							"Field " + fieldId + " is not registered on entity " + entity.name() + "."));
		}
		return entity.effectiveFields().stream().filter(x -> x.clrType() == String.class && x.column() != null)
				.findFirst().orElseThrow(() -> new IllegalStateException(
						"Entity " + entity.name() + " has no string field available for approximate retrieval."));
	}

	private String fuzzySql(EntityMetadata entity, FieldMetadata field) {
		String table = quoteStorage(entity.effectiveStorageName());
		String column = quoteIdentifier(getColumnName(entity, field));
		String identity = quoteIdentifier(getPrimaryKeyName(entity));
		return "SELECT " + identity + ", similarity(" + column + ", ?) AS score " + "FROM " + table + " WHERE " + column
				+ " % ? ORDER BY score DESC LIMIT ?";
	}

	private String fullTextSql(EntityMetadata entity, FieldMetadata field) {
		String table = quoteStorage(entity.effectiveStorageName());
		String column = quoteIdentifier(getColumnName(entity, field));
		String identity = quoteIdentifier(getPrimaryKeyName(entity));
		String config = quoteLiteral(options.fullTextConfiguration());
		return "SELECT " + identity + ", ts_rank_cd(to_tsvector(" + config + ", COALESCE(" + column + "::text, '')), "
				+ "websearch_to_tsquery(" + config + ", ?)) AS score FROM " + table + " WHERE to_tsvector(" + config
				+ ", COALESCE(" + column + "::text, '')) @@ " + "websearch_to_tsquery(" + config
				+ ", ?) ORDER BY score DESC LIMIT ?";
	}

	private String searchSql(EntityMetadata entity, FieldMetadata field) {
		String table = quoteStorage(entity.effectiveStorageName());
		String column = quoteIdentifier(getColumnName(entity, field));
		String identity = quoteIdentifier(getPrimaryKeyName(entity));
		return "SELECT " + identity + ", pdb.score(" + identity + ") AS score FROM " + table + " WHERE " + column
				+ " ||| ? ORDER BY score DESC LIMIT ?";
	}

	private String getColumnName(EntityMetadata entity, FieldMetadata field) {
		if (field.column() == null)
			throw new IllegalStateException("Field " + field.name() + " has no storage column.");
		return entity.columns().stream().filter(x -> x.id().equals(field.column().columnId())).findFirst()
				.orElseThrow(() -> new IllegalStateException("Column " + field.column().columnId() + " for field "
						+ field.name() + " is not registered on entity " + entity.name() + "."))
				.effectiveStorageName();
	}

	private String getPrimaryKeyName(EntityMetadata entity) {
		if (entity.primaryKey() == null)
			throw new IllegalStateException("Entity " + entity.name() + " has no primary key metadata.");
		return entity.columns().stream().filter(x -> x.id().equals(entity.primaryKey().columnId())).findFirst()
				.orElseThrow(() -> new IllegalStateException("Primary key column " + entity.primaryKey().columnId()
						+ " is not registered on entity " + entity.name() + "."))
				.effectiveStorageName();
	}

	private static void requireEnabled(boolean enabled, RetrievalStrategy strategy) {
		if (!enabled)
			throw new UnsupportedOperationException("PostgreSQL retrieval strategy " + strategy + " is disabled.");
	}

	private static String quoteStorage(String value) {
		String[] parts = value.split("\\.");
		if (parts.length == 0)
			throw new IllegalArgumentException("Storage name cannot be empty.");
		return Arrays.stream(parts).filter(x -> !x.isBlank()).map(PostgresRetrievalCandidateSource::quoteIdentifier)
				.reduce((a, b) -> a + "." + b).orElseThrow();
	}

	private static String quoteIdentifier(String value) {
		return "\"" + value.replace("\"", "\"\"") + "\"";
	}

	private static String quoteLiteral(String value) {
		return "'" + value.replace("'", "''") + "'";
	}

	private static String formatScore(double score) {
		return String.format(Locale.ROOT, "%.4f", score);
	}
}
